src/libcore/hint.rs

```diff
@@ -104,11 +104,14 @@ pub fn spin_loop() {
     }
 }
 
-/// A function that is opaque to the optimizer, to allow benchmarks to
-/// pretend to use outputs to assist in avoiding dead-code
-/// elimination.
+/// An identity function that hints the compiler to be maximally pessimistic about what
+/// `black_box` could do.
 ///
-/// This function is a no-op, and does not even read from `dummy`.
+/// The compiler is encouraged to assume that `black_box` can use `x` in any possible valid way
+/// that Rust code is allowed to without introducing undefined behavior in the calling code. This
+/// property makes `black_box` useful for writing code in which certain optimizations are not
+/// desired. However, this also makes `black_box` too unreliable when disabling these optimizations
+/// is required for correctness.
```

Comment:
It would probably be good to include an example using `#[bench]` specifically.

Reply:
I'd rather punt that to some follow up PR. If anything I'd require opening a tracking issue.

Reply:
I agree. I'm running out of steam for `black_box` :)

Reply:
Sure :)